src/Microsoft.ML.Tokenizers/Model/SentencePieceBpe.cs

```diff
@@ -0,0 +1,952 @@
+
+            return new Tokenizer(
+                        new SentencePieceBpe(modelProto, addBeginOfSentence, addEndOfSentence),
+                        SentencePiecePreTokenizer.Instance,
+                        normalizer);
+        }
+
+        /// <summary>
+        /// Specifies whether the model will do a byte fallback when it encounters unknown tokens during the encoding process.
+        /// </summary>
+        public bool ByteFallback { get; }
+
+        /// <summary>
+        /// Indicate emitting the prefix character U+2581 at the beginning of sentence token during the normalization and encoding.
+        /// </summary>
+        public bool AddDummyPrefix { get; }
+
+        /// <summary>
+        /// Indicate if the spaces should be replaced with character U+2581 during the normalization and encoding.
+        /// </summary>
+        public bool EscapeWhiteSpaces { get; }
+
+        /// <summary>
+        /// Indicate emitting the character U+2581 at the end of the last sentence token instead beginning of sentence token during the normalization and encoding.
+        /// </summary>
+        public bool TreatWhitespaceAsSuffix { get; }
+
+        /// <summary>
+        /// Indicate emitting the beginning of sentence token during the encoding.
+        /// </summary>
+        public bool AddBeginningOfSentence { get; }
+
+        /// <summary>
+        /// Indicate emitting the end of sentence token during the encoding.
+        /// </summary>
+        public bool AddEndOfSentence { get; }
+
+        /// <summary>
+        /// The beginning of sentence token.
+        /// </summary>
+        public string BeginningOfSentenceToken { get; }
+
+        /// <summary>
+        /// The end of sentence token.
+        /// </summary>
+        public string EndOfSentenceToken { get; }
+
+        /// <summary>
+        /// The unknown token.
+        /// </summary>
+        public string UnknownToken { get; }
+
+        /// <summary>
+        /// The id of the beginning of sentence token.
+        /// </summary>
+        public int BeginningOfSentenceId { get; }
+
+        /// <summary>
+        /// The id of the end of sentence token.
+        /// </summary>
+        public int EndOfSentenceId { get; }
+
+        /// <summary>
+        /// The id of the unknown token.
+        /// </summary>
+        public int UnknownId { get; }
+
+        /// <summary>
+        /// The vocabulary of the model.
+        /// </summary>
+        public Dictionary<string, int> Vocab
+        {
+            get
+            {
+                if (_publicVocab is null)
+                {
+                    var vocab = new Dictionary<string, int>();
+                    foreach (var item in _vocab)
+                    {
+                        vocab.Add(item.Key.ToString(), item.Value.Id);
+                    }
+
+                    _publicVocab = vocab;
+                }
+                return _publicVocab;
```

Comment:
I assume that even though this is an instance member, we expect instances to be published to a shared location and then for it to be safe for multiple instances to read properties? If so, we should probably sprinkle a bit of synchronization here to ensure that this only ever publishes a single instance:
```suggestion
                Dictionary<string, int> publicVocab = Volatile.Read(ref _publicVocab);
                if (publicVocab is null)
                {
                    var vocab = new Dictionary<string, int>();
                    foreach (var item in _vocab)
                    {
                        vocab.Add(item.Key.ToString(), item.Value.Id);
                    }

                    Interlocked.CompareExchange(ref _publicVocab, vocab, null);
                    publicVocab = _publicVocab;
                }

                return publicVocab ;
```